### DISCUSSION TITLE
Show which NIC is being monitored, and provide the possibility to choose which NICs to monitor

### DISCUSSION THREAD

**Discussion by Matthias-vdE** [henrygd/beszel#94] [Ideas]
For example, my server is connected with two network interface adapters, eno0 and eno1. Both of them are configured and connected via LACP, creating a virtual bond0 interface. Starting a VPN creates yet another virtual interface.

Right now it's not exactly clear which of the NICs is being monitored.

**Comment by henrygd** [comment]
The agent logs the interfaces it uses when it starts up, so look there if you're curious.

There's a very basic check to exclude some interfaces, but it's not perfect, so I'm planning to add an env var to whitelist interfaces at some point.

**Comment by Matthias-vdE** [comment]
@henrygd 

After migrating from my existing Ubuntu to a new Debian server, the configuration for LACP is a bit different when working with VLANS. My server has 2 NICs, enp1s0 and enp2s0. Together they form a bond0 interface, but this interface is actually sub-interfaced as bond0.30. This is the only interface that has an IP address (besides lo0 of course0.

The beszel-agent log detects the first physical, and the bond0 interface:

```
2024-09-05T19:05:02.787713665Z 2024/09/05 19:05:02 Detected network interface: enp1s0 (3299 recv, 1720 sent)

2024-09-05T19:05:02.787741173Z 2024/09/05 19:05:02 Detected network interface: bond0 (3299 recv, 1720 sent)
```

But the bandwidth graph remains at 0 at all times, because traffic is actually going via bond0.30.

**Reply by henrygd** [reply]
The interfaces and stats are pulled from `/proc/net/dev`.

Is bond0.30 included if you run `cat /proc/net/dev`?

**Reply by Matthias-vdE** [reply]
Yes, all interfaces are shown. Including both physical interfaces enp1s0 and enp2s0, as well as bond0 and bond0.30.

I just had a look at the network stats using bmon. And it looks like bond0 is actually sending/receiving data, as well as the physical interface enp2s0 (and virtual interface bond0.30 for that matter).

enp1s0 is not doing very much, as it seems my LACP config is preferring enp2s0. So I assume the graph being shown for bandwidth is pulling data from enp1s0 rather than bond0 because the individual docker network traffic is showing just fine:

![image](https://github.com/user-attachments/assets/ec9270c9-8f94-4b52-988e-363cbb5ded3b)

**Reply by henrygd** [reply]
Can you paste or attach your output from `cat /proc/net/dev`? All of it or just the relevant lines is fine.

The only thing Beszel does is parse that output every minute and calculate the difference since the last check.

If you're using docker, you're using host network mode for the agent, right?

**Comment by Matthias-vdE** [comment]
```
root@debian-server:/# cat /proc/net/dev
Inter-|   Receive                                                |  Transmit
 face |bytes    packets errs drop fifo frame compressed multicast|bytes    packets errs drop fifo colls carrier compressed
    lo: 265982492  458797    0    0    0     0          0         0 265982492  458797    0    0    0     0       0          0
enp1s0: 127342188  524977    0    0    0     0          0    255090 1217933862 1605468    0    0    0     0       0          0
enp2s0: 89084849563 134353877    0    0    0     0          0    255048 6474195771 44054358    0    0    0     0       0          0
 bond0: 89212191751 134878854    0 42426    0     0          0    510138 7692129633 45659826    0    6    0     0       0          0
bond0.30: 85000396793 91841791    0 8485    0     0          0         0 7317927719 42427642    0    3    0     0       0          0

```

And after about one minute:

```
root@debian-server:/# cat /proc/net/dev
Inter-|   Receive                                                |  Transmit
 face |bytes    packets errs drop fifo frame compressed multicast|bytes    packets errs drop fifo colls carrier compressed
    lo: 265995028  458874    0    0    0     0          0         0 265995028  458874    0    0    0     0       0          0
enp1s0: 127352158  525064    0    0    0     0          0    255154 1218587830 1606033    0    0    0     0       0          0
enp2s0: 89123723774 134413431    0    0    0     0          0    255113 6475559011 44073079    0    0    0     0       0          0
 bond0: 89251075932 134938495    0 42436    0     0          0    510267 7694146841 45679112    0    6    0     0       0          0
bond0.30: 85037464194 91883070    0 8487    0     0          0         0 7319834789 42446433    0    3    0     0       0          0
```


Yes the agent is running on network_mode: host

**Reply by henrygd** [reply]
Looks like the agent parses that correctly for me.

```bash
2024/09/08 16:37:11 Detected network interface: enp1s0 (127342188 recv, 1217933862 sent)
2024/09/08 16:37:11 Detected network interface: enp2s0 (89084849563 recv, 6474195771 sent)
2024/09/08 16:37:11 Detected network interface: bond0 (89212191751 recv, 7692129633 sent)
2024/09/08 16:37:11 Detected network interface: bond0.30 (85000396793 recv, 7317927719 sent)
```

Perhaps when the agent was started, only `enp1s0` and `bond0` existed in `/proc/net/dev`? They have very little data transfer in your agent log, so I'm assuming the agent has been running since the machine last started.

Try restarting the agent to see if it picks up the other interfaces this time.

Is this a new issue since updating to the latest version? I did make some changes to the way the interfaces are enumerated to fix a different issue.

Also, even if you get this working I think it will overcount the transfer because it's going to sum up the bytes of all of your interfaces. I will try to get the whitelist env var out in the next release.

**Reply by Matthias-vdE** [reply]
After restarting the agent, I now see the same in the logs:

```
2024-09-09T06:39:32.062086334Z 2024/09/09 06:39:32 Detected network interface: enp1s0 (144888150 recv, 1254877773 sent)

2024-09-09T06:39:32.062126952Z 2024/09/09 06:39:32 Detected network interface: enp2s0 (110740613344 recv, 6959084560 sent)

2024-09-09T06:39:32.062135496Z 2024/09/09 06:39:32 Detected network interface: bond0 (110885501494 recv, 8213962333 sent)

2024-09-09T06:39:32.062142166Z 2024/09/09 06:39:32 Detected network interface: bond0.30 (105746212592 recv, 7805209125 sent)
```

And the bandwidth graph is correctly populating, thanks!

I realize I may have a bit of an exotic setup :) Hence the request to manually specify which NIC to monitor.

**Comment by henrygd** [comment]
`NICS` environment variable has been added in 0.4.0, so you can pass a comma separated whitelist of interfaces to monitor.

Let me know if you have any trouble with it :+1:

**Comment by Matthias-vdE** [comment]
@henrygd 

![image](https://github.com/user-attachments/assets/f04d8972-5fb3-4271-8fa9-3ce5cfb21441)

Looking good! I only whitelisted my bond0 interface, so the graph is even more correct now :) Before: around 1,5Mb/s. After: around 0,5Mb/s.
(Ignoring the peak, was just upgrading some other containers as well)
